modules/@apostrophecms/module/lib/events.js

```diff
@@ -38,18 +39,69 @@ module.exports = function(self) {
         }
       ];
 
-      for (const entry of chain) {
-        const handlers = self.apos.eventHandlers[entry.name] && self.apos.eventHandlers[entry.name][name];
-        if (handlers) {
-          for (const handler of handlers) {
-            const module = self.apos.modules[handler.moduleName];
-            const fn = module.compiledHandlers[entry.name][name][handler.handlerName];
-            // Although we have `self` it can't hurt to
-            // supply the correct `this`
-            await fn.apply(module, args);
+      const telemetry = self.apos.telemetry;
+      const willExit = (name === 'run' && args[0]);
+
+      // Create the "outer" span
+      const moduleName = (self.__meta && self.__meta.name) || 'apostrophe';
+      await telemetry.aposStartActiveSpan(`event:${moduleName}:${name}`, async (spanEmit) => {
+        spanEmit.setAttribute(SemanticAttributes.CODE_FUNCTION, 'emit');
+        spanEmit.setAttribute(SemanticAttributes.CODE_NAMESPACE, moduleName);
+        spanEmit.setAttribute(telemetry.AposAttributes.EVENT_MODULE, moduleName);
+        spanEmit.setAttribute(telemetry.AposAttributes.EVENT_NAME, name);
+
+        if (willExit) {
+          self.apos._onExitQueue.push(() => {
+            console.log(`event:${moduleName}:${name}`);
+            spanEmit.end();
+          });
+        }
+
+        for (const entry of chain) {
+          const handlers = self.apos.eventHandlers[entry.name] && self.apos.eventHandlers[entry.name][name];
+          if (handlers) {
+            for (const handler of handlers) {
+
+              // Create an active "inner" span for each handler using the parent as a context
+              const spanName = `handler:${handler.moduleName}:${handler.handlerName}`;
```

Comment:
handler names are not guaranteed to be unique outside of a given event name so please include the event name here as well.

```
`event:${moduleName}:${name}:handler:${handler.moduleName}:${handler.handlerName}`
```